package/MDAnalysis/lib/nsgrid.pyx

```diff
@@ -424,32 +429,11 @@ cdef class FastNS(object):
         xyz[1] = <int> ((coord[1] - coord[2] * self.cellsize[YZ]) / self.cellsize[YY])
         xyz[0] = <int> ((coord[0] - coord[1] * self.cellsize[XY]
                          - coord[2] * self.cellsize[XZ]) / self.cellsize[XX])
-
-    cdef void coordintoprimarycell(self, float* coord) nogil:
-        cdef float adj_cy, adj_cx
-
-        while (coord[2] >= self.triclinic_dimensions[ZZ]):
-            coord[2] -= self.triclinic_dimensions[ZZ]
-            coord[1] -= self.triclinic_dimensions[YZ]
-            coord[0] -= self.triclinic_dimensions[XZ]
-        while (coord[2] < 0):
-            coord[2] += self.triclinic_dimensions[ZZ]
-            coord[1] += self.triclinic_dimensions[YZ]
-            coord[0] += self.triclinic_dimensions[XZ]
-
-        adj_cy = coord[2] * self.cellsize[YZ]
-        while ((coord[1] - adj_cy) >= self.triclinic_dimensions[YY]):
-            coord[1] -= self.triclinic_dimensions[YY]
-            coord[0] -= self.triclinic_dimensions[XY]
-        while ((coord[1] - adj_cy) < 0):
-            coord[1] += self.triclinic_dimensions[YY]
-            coord[0] += self.triclinic_dimensions[XY]
-
-        adj_cx = coord[2] * self.cellsize[XZ] + coord[1] * self.cellsize[XY]
-        while ((coord[0] - adj_cx) >= self.triclinic_dimensions[XX]):
-            coord[0] -= self.triclinic_dimensions[XX]
-        while ((coord[0] - adj_cx) < 0):
-            coord[0] += self.triclinic_dimensions[XX]
+        # Make sure cell coordinate indices are within the primary unit cell
+        # (better safe than sorry):
+        xyz[0] %= self.ncells[0]
```

Comment:
It should be possible to not need this right? I think the wrapping we're importing has a slightly different style so the above code requires this.

Reply:
I thought this through and the answer is yes and no:
* I wrote the code in `_ortho_pbc()` so that folded coordinates <img src="https://render.githubusercontent.com/render/math?math=r_i\,, i\in\{x,y,z\}"> will always lie within the simulation box so that the condition <img src="https://render.githubusercontent.com/render/math?math=r_i\in[0, L_i)"> is numerically *guaranteed* to be fulfilled (you probably remember because you reviewed the respective PR back then). So, yes, for orthorhombic systems the modulo operations are entirely redundant.
* I had the same in mind at the time I implemented the corresponding operations in `_triclinic_pbc()`, but it turns out that for triclinic boxes, the same guarantee cannot be given for all dimensions simultaneously (at least I think not without seriously impacting performance). The reason for this is that in systems with more than one non-orthogonal box angle, shifting coordinates by one box vector can result in changing more than one dimension of these coordinates. Thus, it can happen that you end up in a situation where eliminating the error in one dimension re-introduces the same problem in another dimension, and vice versa. Nevertheless, the current implementation of `_triclinic_pbc()` at least minimizes this error and guarantees to eliminate it in at least one dimension. Anyway, in the end, the modulo operations are still required!

Reply:
I should probably change the code comment from "better safe than sorry" to "this is strictly required for triclinic boxes".

Reply:
Not sure I follow here. We represent triclinic boxes as a 3x3 matrix which is upper triangular sparse, i.e. `[lx,0,0], [xy,ly,0], [xz,yz,lz]`.  If we do z first, then the remaining 2 dimensions can't affect z, so `xyz[2]` is fixed and won't be affected by `xyz[0/1]`.  Were you getting failures without the modulo operations?

Reply:
I got failures without modulo operations when using `coordintoprimarycell()` because with that function it definitely happens that coordinates can lie outside of the box. With `_triclinic_pbc()` I did not observe any failures without the modulo operations but I left them in as a precaution. After all, they won't affect the correctness of the results and their performance impact is currently not even measurable. I'll have to check the code of `_triclinic_pbc()` once again before I can give a definite statement on whether the modulos are needed.

Reply:
Ok I got your point, and it's exactly what we're doing in `_triclinic_pbc()`. You're right that by working our way up through the matrix from the z-row to the x-row, operations involving the y-row won't affect the z-component and operations involving the x-row won't affect the y- and z-components. From that point of view we should be safe here.
However, I'm still concerned because `_triclinic_pbc()` uses mixed-precision in the sense that the single-precision coordinates are checked to lie inside the box by comparing with the respective upper and lower bounds of the box, and that comparison is done in double precision. In the `FastNS` implementation, however, determining the cell xyz indices is done in single precision, which (I suppose) might lead to different results if coordinates are close to any of the box surfaces.
So I'd still opt for leaving the modulo operations in there as they really won't hurt but instead guarantee that the algorithm works correctly.